README.md
# JdbcRunner

JdbcRunner is a stress testing tool for various RDBMSs.
You can easily create a test scenario with JavaScript and run it in multi-threaded environments.
Additionaly, JdbcRunner bundles some test kits for Oracle Database, MySQL and PostgreSQL so you can use them for benchmarks.

- Tiny sysbench - Port of [sysbench](https://github.com/akopytov/sysbench) OLTP benchmark
- Tiny TPC-B - Simplified implementation of [TPC-B](http://www.tpc.org/tpcb/default.asp)
- Tiny TPC-C - Simplified implementation of [TPC-C](http://www.tpc.org/tpcc/default.asp)

Please see the documents "docs_ja/index.html".
Please see [the website](https://dbstudy.info/jdbcrunner/) for details and [GitHub](https://github.com/sh2/jdbcrunner) for the source code.

## Usage

```
java JR

usage: java JR <script> [options]
-autoCommit <arg>     オートコミットモードを有効化または無効化します (デフォルト : true (有効))
-connLifeTime <arg>   コネクションの最大許容生存時間をミリ秒単位で指定します (デフォルト: ０(無限))
-connPoolSize <arg>   コネクションプールの物理接続数を指定します (デフォルト : nAgents)
-debug                デバッグモードを有効にします (デフォルト : false)
-jdbcDriver <arg>     JDBCドライバのクラス名を指定します (デフォルト : (なし))
-jdbcPass <arg>       データベースユーザのパスワードを指定します (デフォルト : (なし))
-jdbcUrl <arg>        JDBC接続URLを指定します (デフォルト :
                      jdbc:mysql://localhost:3306/test?useSSL=false&allowPublicK
                      eyRetrieval=true)
-jdbcUser <arg>       データベースのユーザ名を指定します (デフォルト : (なし))
-logDir <arg>         ログの出力先ディレクトリを指定します (デフォルト : .)
-measurementTime <arg>測定時間[sec]を指定します (デフォルト : 60)
-nAgents <arg>        エージェント数を指定します (デフォルト : 1)
-param0 <arg>         変数param0に値を設定します
-param1 <arg>         変数param1に値を設定します
-param2 <arg>         変数param2に値を設定します
-param3 <arg>         変数param3に値を設定します
-param4 <arg>         変数param4に値を設定します
-param5 <arg>         変数param5に値を設定します
-param6 <arg>         変数param6に値を設定します
-param7 <arg>         変数param7に値を設定します
-param8 <arg>         変数param8に値を設定します
-param9 <arg>         変数param9に値を設定します
-scriptCharset <arg>  スクリプトの文字セットを指定します
-sleepTime <arg>      トランザクションごとのスリープ時間[msec]を指定します (デフォルト : 0)
-stmtCacheSize <arg>  コネクションあたりの文キャッシュ数を指定します (デフォルト : 10)
-throttle <arg>       スループットの上限値[tps]を指定します (デフォルト : 0 (無制限))
-trace                トレースモードを有効にします (デフォルト : false)
-warmupTime <arg>     測定前にあらかじめ負荷をかけておく時間[sec]を指定します (デフォルト : 10)
```


## Tutorial

Let's run the TPC-C benchmark on MySQL.
First, create a database and a user.

    shell> mysql -u root -p
    sql> CREATE DATABASE tpcc;
    sql> CREATE USER tpcc@'%' IDENTIFIED BY 'tpcc';
    sql> GRANT ALL PRIVILEGES ON tpcc.* TO tpcc@'%';

Then, build the package using ant
    shell> ant

    Buildfile: /yourpath/jdbcrunner/build.xml

    deletedir:
      [delete] Deleting directory /yourpath/jdbcrunner/classes

    makedir:
        [mkdir] Created dir: /yourpath/jdbcrunner/classes

    compile:
        [javac] Compiling 14 source files to /yourpath/jdbcrunner/classes
        [copy] Copying 2 files to /yourpath/jdbcrunner/classes

    create_run_jar:
          [jar] Building jar: /yourpath/jdbcrunner/jdbcrunner-1.4.jar
          [jar] META-INF/LICENSE.txt already added, skipping
          [jar] META-INF/NOTICE.txt already added, skipping
          [jar] META-INF/LICENSE.txt already added, skipping
          [jar] META-INF/NOTICE.txt already added, skipping
          [jar] META-INF/LICENSE.txt already added, skipping
          [jar] META-INF/NOTICE.txt already added, skipping
          [jar] META-INF/services/java.sql.Driver already added, skipping
          [jar] META-INF/LICENSE already added, skipping
          [jar] META-INF/services/java.sql.Driver already added, skipping
          [jar] Building jar: /yourpath/jdbcrunner/jdbcrunner-1.4-nojdbc.jar
          [jar] META-INF/LICENSE.txt already added, skipping
          [jar] META-INF/NOTICE.txt already added, skipping
          [jar] META-INF/LICENSE.txt already added, skipping
          [jar] META-INF/NOTICE.txt already added, skipping
          [jar] META-INF/LICENSE.txt already added, skipping
          [jar] META-INF/NOTICE.txt already added, skipping
    BUILD SUCCESSFUL

Next, load the initial data using the tpcc_load.js script.

    shell> export CLASSPATH=jdbcrunner-1.4.jar
    shell> java JR scripts/tpcc_load.js

    15:37:31 [INFO ] > JdbcRunner 1.4
    15:37:31 [INFO ] [Config]
    Program start time   : 20180815-153731
    Script filename      : scripts/tpcc_load.js
    JDBC driver          : -
    JDBC URL             : jdbc:mysql://localhost:3306/tpcc?useSSL=false&allowPublicKeyRetrieval=true&rewriteBatchedStatements=true
    JDBC user            : tpcc
    Load mode            : true
    Number of agents     : 4
    Auto commit          : false
    Debug mode           : false
    Trace mode           : false
    Log directory        : logs
    Parameter 0          : 0
    Parameter 1          : 0
    Parameter 2          : 0
    Parameter 3          : 0
    Parameter 4          : 0
    Parameter 5          : 0
    Parameter 6          : 0
    Parameter 7          : 0
    Parameter 8          : 0
    Parameter 9          : 0
    15:37:32 [INFO ] Tiny TPC-C - data loader
    15:37:32 [INFO ] -param0  : Scale factor (default : 16)
    15:37:32 [INFO ] -nAgents : Parallel loading degree (default : 4)
    15:37:32 [INFO ] Scale factor            : 16
    15:37:32 [INFO ] Parallel loading degree : 4
    15:37:32 [INFO ] Dropping tables ...
    15:37:32 [INFO ] Creating tables ...
    15:37:33 [INFO ] Loading item ...
    15:37:34 [INFO ] item : 10000 / 100000
    15:37:34 [INFO ] item : 20000 / 100000
    15:37:35 [INFO ] item : 30000 / 100000
    ...
    15:43:40 [INFO ] [Agent 1] orders : 30000 / 30000
    15:43:40 [INFO ] [Agent 0] orders : 30000 / 30000
    15:43:40 [INFO ] [Agent 3] orders : 30000 / 30000
    15:43:40 [INFO ] Creating indexes ...
    15:43:46 [INFO ] Analyzing tables ...
    15:43:46 [INFO ] Completed.
    15:43:46 [INFO ] < JdbcRunner SUCCESS

Then run the benchmark with the tpcc.js script.

    shell> java JR scripts/tpcc.js

    15:45:31 [INFO ] > JdbcRunner 1.3
    15:45:31 [INFO ] [Config]
    Program start time   : 20180815-154531
    Script filename      : scripts/tpcc.js
    JDBC driver          : -
    JDBC URL             : jdbc:mysql://localhost:3306/tpcc?useSSL=false&allowPublicKeyRetrieval=true
    JDBC user            : tpcc
    Warmup time          : 300 sec
    Measurement time     : 900 sec
    Number of tx types   : 5
    Number of agents     : 16
    Connection pool size : 16
    Statement cache size : 40
    Auto commit          : false
    Sleep time           : 0,0,0,0,0 msec
    Throttle             : - tps (total)
    Debug mode           : false
    Trace mode           : false
    Log directory        : logs
    Parameter 0          : 0
    Parameter 1          : 0
    Parameter 2          : 0
    Parameter 3          : 0
    Parameter 4          : 0
    Parameter 5          : 0
    Parameter 6          : 0
    Parameter 7          : 0
    Parameter 8          : 0
    Parameter 9          : 0
    15:45:32 [INFO ] Tiny TPC-C
    15:45:33 [INFO ] Scale factor : 16
    15:45:33 [INFO ] tx0 : New-Order transaction
    15:45:33 [INFO ] tx1 : Payment transaction
    15:45:33 [INFO ] tx2 : Order-Status transaction
    15:45:33 [INFO ] tx3 : Delivery transaction
    15:45:33 [INFO ] tx4 : Stock-Level transaction
    15:45:34 [INFO ] [Warmup] -299 sec, 11,18,2,0,1 tps, (11,18,2,0,1 tx)
    15:45:35 [INFO ] [Warmup] -298 sec, 14,21,1,3,3 tps, (25,39,3,3,4 tx)
    15:45:36 [INFO ] [Warmup] -297 sec, 24,21,4,3,1 tps, (49,60,7,6,5 tx)
    ...
    16:05:31 [INFO ] [Progress] 898 sec, 36,41,5,7,5 tps, 41164,41156,4120,4116,4119 tx
    16:05:32 [INFO ] [Progress] 899 sec, 42,47,5,7,5 tps, 41206,41203,4125,4123,4124 tx
    16:05:33 [INFO ] [Progress] 900 sec, 42,58,3,4,5 tps, 41248,41261,4128,4127,4129 tx
    16:05:33 [INFO ] [Total tx count] 41248,41261,4128,4127,4129 tx
    16:05:33 [INFO ] [Throughput] 45.8,45.8,4.6,4.6,4.6 tps
    16:05:33 [INFO ] [Response time (minimum)] 23,20,5,167,9 msec
    16:05:33 [INFO ] [Response time (50%tile)] 212,63,26,429,36 msec
    16:05:33 [INFO ] [Response time (90%tile)] 329,107,43,609,92 msec
    16:05:33 [INFO ] [Response time (95%tile)] 369,140,51,684,114 msec
    16:05:33 [INFO ] [Response time (99%tile)] 461,268,73,843,158 msec
    16:05:33 [INFO ] [Response time (maximum)] 746,797,127,1237,411 msec
    16:05:33 [INFO ] < JdbcRunner SUCCESS


CHANGES.md
# 1.4 (2022-07-26)
* Upgrade Apache Commons DBCP from commons-dbcp-1.4 to commons-dbcp2-2.9.0 
* Add new option `-connLifeTime`:  コネクションの最大許容生存時間をミリ秒単位で指定します (デフォルト: ０(無限))

# 1.3 (2018-08-21)

* ライセンスを三条項BSDライセンスから二条項BSDライセンスに変更しました。([Commit](https://github.com/sh2/jdbcrunner/commit/13cecc665127f9df1bfe2def853cf7b0879ee3a3))
* 動作環境を更新しました。
    - 動作要件をJava SE 6からJava SE 8に変更しました。([Commit](https://github.com/sh2/jdbcrunner/commit/5abfbfdafc16f36605479e93f2b0c558604c09c4))
    - Commons CLIを1.2から1.4に更新しました。([Commit](https://github.com/sh2/jdbcrunner/commit/2ef3c9f264a6b539889b32c4f705e85f80696b73))
    - Commons Poolを1.5.6から1.6に更新しました。([Commit](https://github.com/sh2/jdbcrunner/commit/03befde30ade3738ed3ccb8d3b7386febf7304d9))
    - Log4jを1.2.16から1.2.17に更新しました。([Commit](https://github.com/sh2/jdbcrunner/commit/0bc1f6d1e71d00e18594adff9ec63fb2e74cc907))
    - Rhinoを1.7R3から1.7.9に更新しました。([Commit](https://github.com/sh2/jdbcrunner/commit/f95f8959b53789853a820ca0c376bdf833adfd69))
* MySQL、PostgreSQLのJDBCドライバを組み込まないバイナリも同梱するようにしました。([Commit](https://github.com/sh2/jdbcrunner/commit/c7ba1f6de46bcc3d7c51a015eb5caaedc7906cad))
* ヒストグラム作成シートをLibreOffice版からExcel版に変更しました。([Commit](https://github.com/sh2/jdbcrunner/commit/464506904a3ae02d640f1ff16a575410b2c26c88))
* getDatabaseMajorVersion()、getDatabaseMinorVersion()ファンクションを追加しました。([Commit](https://github.com/sh2/jdbcrunner/commit/c3e20610330038655a113ede4157447cce0e72ae))
* ランダム文字列を生成するためのrandomString()、setRandomStringElements()ファンクションを追加しました。([Commit](https://github.com/sh2/jdbcrunner/commit/b4dbdffaf6fb8eca8770cd990dd000210b3f99c7))
* sysbench MySQLでのデータロード手順を変更し、高速化しました。([Commit](https://github.com/sh2/jdbcrunner/commit/52d14decaa39d3a30cc6e8e32272f637b2adbecc))
* TPC-B MySQLでのデータロード手順を変更し、高速化しました。([Commit](https://github.com/sh2/jdbcrunner/commit/bea1998c61b5216e86242d95d34817a613c75649))([Commit](https://github.com/sh2/jdbcrunner/commit/25f8475c71dbdb5c8711f5e1355fdf91666782f4))
* TPC-CのデータロードをrandomString()ファンクションを用いて高速化しました。([Commit](https://github.com/sh2/jdbcrunner/commit/b4dbdffaf6fb8eca8770cd990dd000210b3f99c7))
* TPC-C MySQLでのデータロード手順を変更し、高速化しました。([Commit](https://github.com/sh2/jdbcrunner/commit/cf2d07cdc652fc6a1689f804e6c53dbb1926f244))([Commit](https://github.com/sh2/jdbcrunner/commit/4100086ca6ae9ae3d41bfa630ef1350a6882e08e))
* TPC-C PaymentトランザクションがPostgreSQLでデッドロックを起こしていたのを軽減しました。([Commit](https://github.com/sh2/jdbcrunner/commit/5668990d0813c3b2be84208a0cc5c50e7496274f))(by [myzkyy](https://github.com/myzkyy))
* TPC-C DeliveryトランザクションのSELECT FOR UPDATE文においてサブクエリを使わないように修正し、より多くのRDBMSで動作するようにしました。([Commit](https://github.com/sh2/jdbcrunner/commit/660226be7e7e2af4dd8d4a605fa9b0db242059b7))
* TPC-C customerテーブルのc_since列に、現在時刻ではなくデータロード開始時刻を入れるように修正しました。([Commit](https://github.com/sh2/jdbcrunner/commit/a48a95d87c008e1454ccb601278b0a6679e4d743))
* IBM Db2用のサンプルスクリプトを追加しました。([Commit](https://github.com/sh2/jdbcrunner/commit/a5076a694cc47a7ec8d5863f1b0ee07e50033cd9))(by [nakunaru](https://github.com/nakunaru))
* Oracle Databaseでテスト終了時にユーザ定義エラーが発生することがあったのを修正しました。([Commit](https://github.com/sh2/jdbcrunner/commit/fa4d5df39888e870367ba69546de9978748ee081))([Commit](https://github.com/sh2/jdbcrunner/commit/5b8714814c0e5501ad25dcccfb35915e94170e23))

# 1.2 (2011-10-11)

* The option -throttle checks busyness of transactions. When throughput is lower than the configured value, JdbcRunner gives up keeping throughput in total and tries to keep it from now on.
* The throughput is calculated down to the first decimal place.
* Tiny TPC-C was updated to version 1.1.
    - JdbcRunner executes transactions in more precise ratio.
    - An USE_NL hint was added to SQL S-01 for Oracle Database.
* A Calc file was added to draw response time histogram.
* Apache Commons Pool was updated to version 1.5.6.
* Apache log4j was updated to version 1.2.16.
* Mozilla Rhino was updated to version 1.7R3.
* MySQL Connector/J was updated to version 5.1.18.
* PostgreSQL JDBC Driver was updated to version 9.1 Build 901.

# 1.1 (2010-05-06)

* Long and double data types were supported for bind variables.
* Null value was supported for bind variables.
* A new function getScriptStackTrace() was added.
* A new test kit Tiny SysBench was added.
* Tiny TPC-B was updated to version 1.1.
    - Foreign key constraints were added.
* A new test kit Tiny TPC-C was added.
* Apache Commons DBCP was updated to version 1.4. DBCP 1.4 now supports JDBC 4.0.
* MySQL Connector/J was updated to version 5.1.12.
* The default value of jdbcDriver was changed to empty string. JDBC 4.0 drivers no longer need this parameter.

# 1.0 (2010-01-10)

* First release.


src/jdbcrunner/Agent.java
package jdbcrunner;

import java.sql.Connection;
import java.sql.SQLException;

/**
 * 負荷シナリオを実行するクラスです。
 * <p>
 * このクラスのインスタンスは、スレッドを1つ起動して負荷シナリオを実行します。
 *
 * @author Sadao Hiratsuka
 */
public class Agent implements Runnable {
	private final int id;
	private final Manager manager;
	private final Config config;
	private final Throttle throttle;
	private final Record record;

	private volatile Thread thread;
	private boolean doBreak;
	private int txType = 0;

	/**
	 * エージェントのID、負荷テストを管理するマネージャ、
	 * 負荷テストの設定を指定してエージェントを構築します。
	 *
	 * @param id
	 *            エージェントのID
	 * @param manager
	 *            負荷テストを管理するマネージャ
	 * @param config
	 *            負荷テストの設定
	 * @param throttle
	 *            スループット制御オブジェクト
	 */
	public Agent(int id, Manager manager, Config config, Throttle throttle) {
		this.id = id;
		this.manager = manager;
		this.config = config;
		this.throttle = throttle;
		this.record = new Record(config);
	}

	/**
	 * スレッドを起動して、負荷シナリオを開始します。
	 */
	public void start() {
		if (thread == null) {
			this.thread = new Thread(this, "agent" + Integer.toString(id)); //$NON-NLS-1$
			thread.start();
		}
	}

	/**
	 * 負荷シナリオを実行します。
	 * <p>
	 * <ol>
	 * <li>{@code init()}ファンクションを1回実行します。
	 * <li>初期化処理の完了ラッチを解除します。
	 * <li>測定の開始ラッチが解除されるのを待ちます。
	 * <li>測定開始時刻をインスタンス変数に登録します。
	 * <li>以下の条件を満たす間{@code run()}
	 * ファンクションを繰り返し実行します。
	 * <ul>
	 * <li>負荷テストが正常に行われている
	 * <li>負荷テストの進捗状況が{@code RUNNING}である
	 * <li>スクリプトの停止フラグが有効になっていない
	 * </ul>
	 * <li>測定の完了ラッチを解除します。
	 * <li>終了処理の開始ラッチが解除されるのを待ちます。
	 * <li>{@code fin()}ファンクションを1回実行します。
	 * <li>終了処理の完了ラッチを解除します。
	 * </ol>
	 * <p>
	 * 途中で例外が発生した場合は、以降の処理は中止してマネージャに例外を通知します。
	 */
	@Override
	public void run() {
		Helper helper = null;

		try {
			helper = new Helper(config, this);

			// 初期化処理を行う
			callInit(helper);
			manager.getInitEndLatch().countDown();

			// 測定を行う
			while (manager.getRunStartLatch().getCount() > 0) {
				try {
					manager.getRunStartLatch().await();
				} catch (InterruptedException e) {
					putMessage(new Message(Message.Level.TRACE,
							Resources.getString("Agent.INTERRUPTED_EXCEPTION"))); //$NON-NLS-1$
				}
			}

			record.setStartTime(manager.getStartTime());

			while (manager.getCondition() && (manager.getStatus() == Manager.Status.RUNNING)
					&& !doBreak) {

				long txBeginTime = 0L;
				long txEndTime = 0L;

				txBeginTime = System.nanoTime();
				callRun(helper);
				txEndTime = System.nanoTime();

				manager.incrementTxCount(txType);
				record.add(txType, txBeginTime, txEndTime);
				sleep();
				throttle(txEndTime);
			}

			manager.getRunEndLatch().countDown();

			// 終了処理を行う
			while (manager.getFinStartLatch().getCount() > 0) {
				try {
					manager.getFinStartLatch().await();
				} catch (InterruptedException e) {
					putMessage(new Message(Message.Level.TRACE,
							Resources.getString("Agent.INTERRUPTED_EXCEPTION"))); //$NON-NLS-1$
				}
			}

			if (manager.getCondition()) {
				callFin(helper);
			}

			manager.getFinEndLatch().countDown();
		} catch (ApplicationException e) {
			putMessage(new Message(Message.Level.ERROR,
					Resources.getString("Agent.EXCEPTION_1") + id //$NON-NLS-1$
							+ Resources.getString("Agent.EXCEPTION_2"), e)); //$NON-NLS-1$
		} catch (Exception e) {
			putMessage(new Message(Message.Level.ERROR,
					Resources.getString("Agent.EXCEPTION_1") + id //$NON-NLS-1$
							+ Resources.getString("Agent.EXCEPTION_2"), e)); //$NON-NLS-1$
		} finally {
			if (helper != null) {
				helper.closeScript();
			}
		}
	}

	/**
	 * エージェントのIDを返します。
	 *
	 * @return エージェントのID
	 */
	public int getId() {
		return id;
	}

	/**
	 * エージェントが実行したトランザクションの記録を返します。
	 *
	 * @return トランザクションの記録
	 */
	public Record getRecord() {
		return record;
	}

	/**
	 * エージェントが起動したスレッドを返します。
	 * <p>
	 * このメソッドはスレッドセーフです。
	 *
	 * @return スレッド。スレッドがまだ起動されていないときは{@code null}
	 */
	public Thread getThread() {
		return thread;
	}

	/**
	 * 測定の停止フラグを立てます。
	 */
	public void setBreak() {
		this.doBreak = true;
	}

	/**
	 * トランザクション種別を設定します。
	 *
	 * @param txType
	 *            トランザクション種別の番号。これは0以上
	 *            {@code Config#getNTxTypes()}
	 *            未満である必要があります
	 * @throws IllegalArgumentException
	 *             {@code txType}が範囲外の場合
	 */
	public void setTxType(int txType) {
		if ((txType >= 0) && (txType < config.getNTxTypes())) {
			this.txType = txType;
		} else {
			throw new IllegalArgumentException(Resources.getString("Agent.TXTYPE_OUT_OF_RANGE") //$NON-NLS-1$
					+ txType);
		}
	}

	/**
	 * マネージャにメッセージを通知します。
	 *
	 * @param message
	 *            メッセージ
	 * @see Manager#putMessage(Message)
	 */
	public void putMessage(Message message) {
		manager.putMessage(message);
	}

	private void callInit(Helper helper) throws ApplicationException {
		Connection connection = null;

		try {
			connection = manager.getConnection();
			helper.setConnection(connection);
			helper.callInit();
		} catch (Exception e) {
			throw new ApplicationException(Resources.getString("Agent.SQL_EXCEPTION"), e); //$NON-NLS-1$
		} finally {
			if (connection != null) {
				try {
					connection.close();
				} catch (SQLException e) {
					// 何もしない
				}
			}
		}
	}

	private boolean callRun(Helper helper) throws ApplicationException {
		Connection connection = null;
		boolean result = false;

		try {
			connection = manager.getConnection();
			helper.setConnection(connection);
			helper.callRun();
			result = true;
		} catch (SQLException e) {
		//} catch (Exception e) {
			// throw new ApplicationException(Resources.getString("Agent.SQL_EXCEPTION"), e); //$NON-NLS-1$
			putMessage(new Message(Message.Level.ERROR,
					Resources.getString("Agent.SQL_EXCEPTION"), e)); //$NON-NLS-1$
			result = false;
			try {
				Thread.sleep(1000);	// 1秒待つ
			} catch (InterruptedException e2) {
				// 何もしない
			}
		} finally {
			if (connection != null) {
				try {
					connection.close();
				} catch (SQLException e) {
					// 何もしない
				}
			}
		}
		return result;
	}

	private void callFin(Helper helper) throws ApplicationException {
		Connection connection = null;

		try {
			connection = manager.getConnection();
			helper.setConnection(connection);
			helper.callFin();
		} catch (SQLException e) {
			throw new ApplicationException(Resources.getString("Agent.SQL_EXCEPTION"), e); //$NON-NLS-1$
		} finally {
			if (connection != null) {
				try {
					connection.close();
				} catch (SQLException e) {
					// 何もしない
				}
			}
		}
	}

	private void sleep() {
		if ((config.getSleepTime(txType) > 0L) && manager.getCondition()
				&& (manager.getStatus() == Manager.Status.RUNNING)) {

			try {
				Thread.sleep(config.getSleepTime(txType));
			} catch (InterruptedException e) {
				putMessage(new Message(Message.Level.TRACE,
						Resources.getString("Agent.INTERRUPTED_EXCEPTION"))); //$NON-NLS-1$
			}
		}
	}

	private void throttle(long txEndTime) {
		if ((config.getThrottle(txType) > 0) && manager.getCondition()
				&& (manager.getStatus() == Manager.Status.RUNNING)) {

			try {
				Thread.sleep(throttle.getSleepTime(txType, txEndTime));
			} catch (InterruptedException e) {
				putMessage(new Message(Message.Level.TRACE,
						Resources.getString("Agent.INTERRUPTED_EXCEPTION"))); //$NON-NLS-1$
			}
		}
	}
}


scripts/sysbench_load.js
/*
 * Tiny sysbench - data loader
 * This script is based on sysbench 0.4.12.
 * https://github.com/akopytov/sysbench
 *
 * [Oracle Database]
 * shell> sqlplus "/ AS SYSDBA"
 * sql> CREATE USER sbtest IDENTIFIED BY sbtest;
 * sql> GRANT CREATE SESSION, CREATE TABLE, CREATE SEQUENCE, CREATE TRIGGER, UNLIMITED TABLESPACE TO sbtest;
 *
 * [MySQL]
 * shell> mysql -u root -p
 * sql> CREATE DATABASE sbtest;
 * sql> CREATE USER sbtest@'%' IDENTIFIED BY 'sbtest';
 * sql> GRANT ALL PRIVILEGES ON sbtest.* TO sbtest@'%';
 *
 * [PostgreSQL]
 * shell> psql -U postgres
 * sql> CREATE DATABASE sbtest TEMPLATE template0 ENCODING 'UTF-8' LC_COLLATE 'C' LC_CTYPE 'C';
 * sql> CREATE USER sbtest PASSWORD 'sbtest';
 *
 * <postgresql.conf>
 * listen_addresses = '*'
 * port = 5432
 *
 * <pg_hba.conf>
 * host all all 0.0.0.0/0 md5
 */

// JdbcRunner settings -----------------------------------------------

// Oracle Database
// var jdbcUrl = "jdbc:oracle:thin:@//localhost:1521/orcl.local";

// MySQL
var jdbcUrl = "jdbc:mysql://localhost:3306/sbtest?useSSL=false&allowPublicKeyRetrieval=true&rewriteBatchedStatements=true";

// PostgreSQL
// var jdbcUrl = "jdbc:postgresql://localhost:5432/sbtest";

var jdbcUser = "sbtest";
var jdbcPass = "sbtest";
var isLoad = true;
var isAutoCommit = false;
var logDir = "logs";
var nAgents = 16;

// Application settings ----------------------------------------------

var BATCH_SIZE = 10000;
var COMMIT_SIZE = 10000;

var oltpTableSize;
var oltpTableCount;

// JdbcRunner functions ----------------------------------------------

function init() {
    var agentId = getId();
    info("init() : agent " + agentId);

    if (getId() == 0) {
        info("Tiny sysbench - data loader");
        info("-nAgents: Number of Agents  (default: 16)");
        info("-param0 : Number of records (default : 10000)");
        info("-param1 : Number of tables (default : 32)");
        info("-param2 : (for TiDB) Create index before insert. 1 or 0 (default : 0)");

        oltpTableSize = param0;
        oltpTableCount = param1;
        createIndexBeforeInsert = param2;

        if (oltpTableSize == 0) {
            oltpTableSize = 10000;
        }
        putData("OLTPTableSize", oltpTableSize);
        if (oltpTableCount == 0) {
            oltpTableCount = 32;
        }
        putData("OLTPTableCount", oltpTableCount);
        if (createIndexBeforeInsert == 0) {
            createIndexBeforeInsert = 0;
        } else {
            createIndexBeforeInsert = 1;
        }
        putData("CreateIndexBeforeInsert", createIndexBeforeInsert);

        info("Number of records : " + oltpTableSize);
        info("Number of tables : " + oltpTableCount);
        info("Create index before insert : " + createIndexBeforeInsert);

        for (var tableId = 1; tableId <= oltpTableCount; tableId++) {
            if (getDatabaseProductName() == "Oracle") {
                dropTableOracle(tableId);
                createTableOracle(tableId);
                if (createIndexBeforeInsert == 1) {
                    createIndexOracle(tableId);
                }
            } else if (getDatabaseProductName() == "MySQL") {
                dropTableMySQL(tableId);
                createTableMySQL(tableId);
                if (createIndexBeforeInsert == 1) {
                    createIndexMySQL(tableId);
                }
            } else if (getDatabaseProductName() == "PostgreSQL") {
                dropTablePostgreSQL(tableId);
                createTablePostgreSQL(tableId);
                if (createIndexBeforeInsert == 1) {
                    createIndexPostgreSQL(tableId);
                }
            } else {
                error(getDatabaseProductName() + " is not supported yet.");
            }
        }
        
        commit();
    }
}

function run() {
    if (!oltpTableSize) {
        oltpTableSize = Number(getData("OLTPTableSize"));
    }
    if (!oltpTableCount) {
        oltpTableCount = Number(getData("OLTPTableCount"));
    }
    var agentId = getId();
    for (var tableId = agentId + 1; tableId <= oltpTableCount; tableId += nAgents) {
        var tableName = "sbtest" + tableId;
        info("Loading " + tableName + "...");
        
        var k = new Array();
        var c = new Array();
        var pad = new Array();
        
        for (var count = 1; count <= oltpTableSize; count++) {
            k.push(0);
            c.push(" ");
            pad.push("qqqqqqqqqqwwwwwwwwwweeeeeeeeeerrrrrrrrrrtttttttttt");
            
            if (count % BATCH_SIZE == 0) {
                executeBatch("INSERT INTO " + tableName + " (k, c, pad) VALUES ($int, $string, $string)",
                    k, c, pad);
                
                k.length = 0;
                c.length = 0;
                pad.length = 0;
                
                if (count % COMMIT_SIZE == 0) {
                    commit();
                    info(tableName + " : " + count + " / " + oltpTableSize);
                }
            }
        }
        
        if (oltpTableSize % COMMIT_SIZE != 0) {
            if (oltpTableSize % BATCH_SIZE != 0) {
                executeBatch("INSERT INTO " + tableName + " (k, c, pad) VALUES ($int, $string, $string)",
                    k, c, pad);
            }
            
            commit();
            info(tableName + " : " + oltpTableSize + " / " + oltpTableSize);
        }
    }
    
    setBreak();
}

function fin() {
    var agentId = getId();
    oltpTableCount = Number(getData("OLTPTableCount"));
    createIndexBeforeInsert = Number(getData("CreateIndexBeforeInsert"));
    for (var tableId = agentId + 1; tableId <= oltpTableCount; tableId += nAgents) {
        if (getDatabaseProductName() == "Oracle") {
            if (createIndexBeforeInsert == 0) {
                createIndexOracle(tableId);
            }
            gatherStatsOracle();
        } else if (getDatabaseProductName() == "MySQL") {
            if (createIndexBeforeInsert == 0) {
                createIndexMySQL(tableId);
            }
            gatherStatsMySQL(tableId);
        } else if (getDatabaseProductName() == "PostgreSQL") {
            if (createIndexBeforeInsert == 0) {
                createIndexPostgreSQL(tableId);
            }
            gatherStatsPostgreSQL(tableId);
        } else {
            error(getDatabaseProductName() + " is not supported yet.");
        }
    }
    commit();
    info("agent " + agentId + " Completed");
}

// Application functions ---------------------------------------------

function dropTableOracle(tableId) {
    var tableName = "sbtest" + tableId;
    info("Dropping a table " + tableName + " ...");
    
    try {
        execute("DROP TABLE " + tableName + "");
        execute("DROP SEQUENCE " + tableName + "_seq");
    } catch (e) {
        warn(e);
    }
}

function dropTableMySQL(tableId) {
    var tableName = "sbtest" + tableId;
    info("Dropping a table " + tableName + " ...");
    
    try {
        execute("DROP TABLE " + tableName + "");
    } catch (e) {
        warn(e);
        rollback(); // PostgreSQL requires a rollback.
    }
}

function dropTablePostgreSQL(tableId) {
    dropTableMySQL(tableId);
}

function createTableOracle(tableId) {
    var tableName = "sbtest" + tableId;
    info("Creating a " + tableName + " ...");
    
    execute("CREATE TABLE " + tableName + " ("
        + "id NUMBER, "
        + "k NUMBER DEFAULT 0 NOT NULL, "
        + "c CHAR(120) DEFAULT '' NOT NULL, "
        + "pad CHAR(60) DEFAULT '' NOT NULL)");
    
    execute("CREATE SEQUENCE " + tableName + "_seq");
    
    var statement;
    
    try {
        statement = takeConnection().createStatement();
        
        statement.executeUpdate("CREATE TRIGGER " + tableName + "_trig "
            + "BEFORE INSERT ON " + tableName + " FOR EACH ROW "
            + "BEGIN "
                + "IF :NEW.id IS NULL THEN "
                    + "SELECT " + tableName + "_seq.NEXTVAL INTO :NEW.id FROM DUAL; "
                + "END IF; "
            + "END;");
    } finally {
        try {
            statement.close();
        } catch (e) {
            warn(e);
        }
    }
}

function createTableMySQL(tableId) {
    var tableName = "sbtest" + tableId;
    info("Creating a " + tableName + " ...");
    
    execute("CREATE TABLE " + tableName + " ("
        + "id INT UNSIGNED PRIMARY KEY AUTO_INCREMENT, "
        + "k INT UNSIGNED DEFAULT 0 NOT NULL, "
        + "c CHAR(120) DEFAULT '' NOT NULL, "
        + "pad CHAR(60) DEFAULT '' NOT NULL) "
        + "ENGINE = InnoDB");
}

function createTablePostgreSQL(tableId) {
    var tableName = "sbtest" + tableId;
    info("Creating a " + tableName + " ...");
    
    execute("CREATE TABLE " + tableName + " ("
        + "id SERIAL, "
        + "k INTEGER DEFAULT 0 NOT NULL, "
        + "c CHAR(120) DEFAULT '' NOT NULL, "
        + "pad CHAR(60) DEFAULT '' NOT NULL)");
}

function createIndexOracle(tableId) {
    var tableName = "sbtest" + tableId;
    info("Creating an index for " + tableName + " ...");
    
    execute("ALTER TABLE " + tableName + " ADD CONSTRAINT " + tableName + "_pk PRIMARY KEY (id)");
    execute("CREATE INDEX " + tableName + "_ix1 ON " + tableName + " (k)");
}

function createIndexMySQL(tableId) {
    var tableName = "sbtest" + tableId;
    info("Creating an index for " + tableName + " ...");
    
    execute("ALTER TABLE " + tableName + " ADD KEY " + tableName + "_ix1 (k)");
}

function createIndexPostgreSQL(tableId) {
    createIndexOracle(tableId);
}

function gatherStatsOracle() {
    info("Analyzing a table ...");
    
    execute("BEGIN DBMS_STATS.GATHER_SCHEMA_STATS(ownname => NULL); END;");
}

function gatherStatsMySQL(tableId) {
    var tableName = "sbtest" + tableId;
    info("Analyzing a table " + tableName + " ...")
    
    execute("ANALYZE TABLE " + tableName);
}

function gatherStatsPostgreSQL(tableId) {
    var tableName = "sbtest" + tableId;
    info("Vacuuming and analyzing a table " + tableName + " ...");
    
    takeConnection().setAutoCommit(true);
    execute("VACUUM ANALYZE " + tableName);
    takeConnection().setAutoCommit(false);
}
